src/goto-programs/destructor_tree.cpp

```diff
@@ -0,0 +1,108 @@
+#include "destructor_tree.h"
+
+/// Adds a destructor to the current stack, either attaching itself to the
+/// closest previous node, or an appropriate rewind target.
+void destructor_treet::add(const codet &destructor)
+{
+  // If we have a stored rewind target, attach node to that then wipe rewind.
+  auto previous_node = get_current_node();
+  auto new_node = destruction_graph.add_node(destructor);
+  if(rewind_target)
+  {
+    auto rewind_target_index = *rewind_target;
+    rewind_target = {};
+    if(!destruction_graph.has_edge(rewind_target_index, new_node))
+      destruction_graph.add_edge(rewind_target_index, new_node);
+  }
+  else
+  {
+    destruction_graph.add_edge(previous_node, new_node);
+  }
+}
+
+/// Fetches the destructor value for the passed-in node identity.
+optionalt<codet> &destructor_treet::get(std::size_t identity)
+{
+  return destruction_graph[identity].destructor_value;
+}
+
+/// Finds the node two branches intersect on and then returns it.
+/// This should be used when you want to find out what parts of the two
+/// branches are common to each other.
+const intersection_resultt destructor_treet::get_intersection_node(
+  optionalt<std::size_t> left_optional,
```

Comment:
Add some typedef so that you can use `node_indext` instead of the very generic `std::size_t`. That would help make clear that these are (optional) node indices.